src/wallet/coin_selection.rs

```diff
@@ -138,6 +145,98 @@ impl CoinSelectionResult {
+/// * Timing cost
+/// * Creation cost
+/// > waste = timing_cost + creation_cost
+///
+/// **Timing cost** is the cost associated to the current fee rate and some long term fee rate used
+/// as a treshold to consolidate UTXOs.
+/// > timing_cost = txin_size * current_fee_rate - txin_size * long_term_fee_rate
+///
+/// Timing cost can be negative if the `current_fee_rate` is cheaper than the `long_term_fee_rate`,
+/// or zero if they are equal.
+///
+/// **Creation cost** is the cost associated to the surplus of coins besides the transaction amount
+/// and transaction fees. It can happen in the form of a change output or in the form of excess
+/// fees paid to the miner.
+///
+/// Change cost is derived from the cost of adding the extra output to the transaction and spending
+/// that output in the future.
+/// > change_cost = current_fee_rate * change_output_size + long_term_feerate * change_spend_size
+///
+/// Excess happens when there is not change, and the surplus of coins is spend as part of the fees
+/// to the miner:
+/// > excess = tx_total_value - tx_fees - target
+///
+/// Where _target_ is the desired amount to send.
+///
+/// Creation cost can be zero if there is a perfect match as result of the coin selection
+/// algorithm.
+///
+/// So, waste can be zero or negative if the creation cost is zero and the timing
+/// cost is less than or equal to zero.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Waste(pub i64);
+
+const LONG_TERM_FEE_RATE: FeeRate = FeeRate::from_sat_per_vb(10.0);
+
+impl Waste {
+    /// Calculate the amount of waste for the given coin selection
+    ///
+    /// - `selected`: the selected transaction inputs
+    /// - `cost_of_change`: the cost of creating change and spending it in the future
+    /// - `target`: the amount in satoshis of the aimed transaction output
+    /// - `fee_rate`: fee rate to use
+    pub fn calculate(
+        selected: &[WeightedUtxo],
+        cost_of_change: Option<u64>,
+        target: u64,
+        fee_rate: FeeRate,
+    ) -> Result<Waste, Error> {
+        // Always consider the cost of spending an input now vs in the future.
+        let utxo_groups: Vec<_> = selected
+            .iter()
+            .map(|u| OutputGroup::new(u, fee_rate))
+            .collect();
+
+        // If fee_rate < LONG_TERM_FEE_RATE, timing cost can be negative
+        let timing_cost: i64 = utxo_groups.iter().fold(0, |acc, utxo| {
+            let fee: i64 = utxo.fee as i64;
+            let long_term_fee: i64 = LONG_TERM_FEE_RATE
+                .fee_wu(TXIN_BASE_WEIGHT + utxo.weighted_utxo.satisfaction_weight)
+                as i64;
+
+            acc + fee - long_term_fee
+        });
+
+        let selected_effective_value: i64 = utxo_groups
+            .iter()
+            .fold(0, |acc, utxo| acc + utxo.effective_value);
+
+        let creation_cost = match cost_of_change {
+            Some(change_cost) => {
```

Comment:
You could perhaps detect whether the remainder after paying for the fixed transaction parts is sufficient to create a change, rather than making the caller responsible for knowing whether there will be change. A possible lower bound of satoshis in the remainder for toggling change creation could be `change_output_size×feerate + dust_threshold`.

Reply:
I think that it's desirable to do it, but it breaks the meaning of metric.
IMHO, this metric should only measure and provide a single outcome through
its result.
The intention is to use it as a tie breaker between the different
coin selection algorithms.
So, it will be computed after the coin selection algorithm. At that moment,
they should already have selected the UTXOs and decide whether there will be
change or not.
I think that we can reconsider this after it has been integrated with the
CoinSelectionAlgorithms.

Reply:
Right, if you use the heuristic to compare transaction candidates after coin selection, you'll know already whether a change is created. The automatic detection could be useful if you use it during coin selection to evaluate candidates of an algorithm that can optionally create change.